lib/Dialect/Arc/Transforms/SimplifyVariadicOps.cpp

```diff
@@ -0,0 +1,127 @@
+//===- SimplifyVariadicOps.cpp --------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-simplify-variadic-ops"
+
+using namespace mlir;
+using namespace circt;
+using namespace arc;
+using namespace hw;
+
+namespace {
+struct SimplifyVariadicOpsPass
+    : public SimplifyVariadicOpsBase<SimplifyVariadicOpsPass> {
+  SimplifyVariadicOpsPass() = default;
+  SimplifyVariadicOpsPass(const SimplifyVariadicOpsPass &pass)
+      : SimplifyVariadicOpsPass() {}
+
+  void runOnOperation() override;
+  void simplifyOp(Operation *op);
+
+  Statistic numOpsSkippedMultipleBlocks{
+      this, "skipped-multiple-blocks",
+      "Ops skipped due to operands in different blocks"};
+  Statistic numOpsSimplified{this, "simplified",
+                             "Ops simplified into binary ops"};
+  Statistic numOpsCreated{this, "created",
+                          "Ops created as part of simplification"};
+  Statistic numOpsReordered{this, "reordered",
+                            "Ops where simplification reordered operands"};
```

Comment:
nit: can we move these declarations into tablegen?

Reply:
Today I learned: TableGen supports pass statistics 🥳